### PR TITLE
Add support for instance-permission-set in azure

### PR DESCRIPTION
**Pull request by tasdomas**
This PR addresses #559 

The instance-permission-set is parsed as a comma-separated list of ARM resource ids in the form
'/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}'.

### CODE REVIEW

**Comment by dacbd** [comment]
You got https://github.com/iterative/terraform-provider-iterative/issues/235 ✔️ 😉 
#559 would live here: https://github.com/iterative/terraform-provider-iterative/tree/master/task/az

**Comment by tasdomas** [comment]
@dacbd ah, I see - I guess I misunderstood the relation between `tpi` and `task`

**Comment by dacbd** [comment]
> @dacbd ah, I see - I guess I misunderstood the relation between `tpi` and `task`

no worries they both needed to be done at some point 👍

File: task/az/resources/data_source_permission_set.go [hunk 1]
```diff
@@ -3,12 +3,17 @@ package resources
...
+	"regexp"
+	"strings"
 
 	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-30/compute"
 
 	"terraform-provider-iterative/task/az/client"
 )
 
+// validateARMID is a regular expression for validating user-assigned identity ids.
+var validateARMID = regexp.MustCompile("/subscriptions/.*/resourceGroups/.*/providers/Microsoft.ManagedIdentity/userAssignedIdentities/.*")
```
[the remark this answers is not in the document]
Comment (tasdomas) [reply task/az/resources/data_source_permission_set.go:15]: I think it makes sense to do that. But first I need to figure out why the smoke tests are failing

File: task/az/task.go [hunk 2]
```diff
@@ -110,48 +110,59 @@ type Task struct {
...
-	logrus.Info("[8/10] Creating VirtualMachineScaleSet...")
-	if err := t.Resources.VirtualMachineScaleSet.Create(ctx); err != nil {
-		return err
-	}
-	logrus.Info("[9/10] Uploading Directory...")
-	if t.Attributes.Environment.Directory != "" {
-		if err := t.Push(ctx, t.Attributes.Environment.Directory); err != nil {
+	// Define a list of steps to execute. If we adopt this pattern, we
+	// can declare a named type for the step definition.
+	steps := []struct {
```
Comment (DavidGOrtega) [inline task/az/task.go:115]: Is this change needed in this PR? I find it a bit out of the scope

File: iterative/azure/provider.go [hunk 3]
```diff
@@ -360,6 +364,27 @@ func getVMClient(subscriptionID string) (compute.VirtualMachinesClient, error) {
 	return client, err
 }
 
+// getUserAssignedIdentityMap generates a map of user-assigned identity values from a comma-separated
+// list of ARM resource ids.
+func getUserAssignedIdentityMap(identitiesRaw string) map[string]*compute.VirtualMachineIdentityUserAssignedIdentitiesValue {
```
Comment (DavidGOrtega) [inline iterative/azure/provider.go:369]: This code is very similar to 

```
identities := strings.Split(ps.Identifier, ",")
	identityMap := map[string]*compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue{}
	for _, identity := range identities {
		identity = strings.TrimSpace(identity)
		if identity == "" {
			continue
		}
		if !validateARMID.MatchString(identity) {
			return fmt.Errorf("invalid user-assigned identity id: %q", identity)
		}
		identityMap[identity] = &compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue{}
	}
	if len(identityMap) == 0 {
		ps.Resource = nil
		return nil
	}
	ps.Resource = &compute.VirtualMachineScaleSetIdentity{
		UserAssignedIdentities: identityMap,
```

are they different in scope? Could them be using a common function?

Comment (tasdomas) [reply iterative/azure/provider.go:369]: They are pretty much the same thing, but I've not yet found a good place to put common azure-specific functionality. Also, there's the subtle difference that in one place we're creating a map of `*compute.VirtualMachineIdentityUserAssignedIdentitiesValue` and in the other - a map of `*compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue`. 

DRY is a nice principle in theory, in practice it comes with its own cost.

File: task/az/task.go (hunk 2 above)
Comment (DavidGOrtega) [reply task/az/task.go:115]: Agree, but why not follow up in technical debt issue? Just only to not loose the focus when reviewing it.

File: iterative/azure/provider.go (hunk 3 above)
Comment (DavidGOrtega) [reply iterative/azure/provider.go:369]: Im not that picky but fi there is an easy way to just call a generic map creator would be awesome.

File: task/az/task.go (hunk 2 above)
Comment (dacbd) [reply task/az/task.go:115]: it looks like https://github.com/iterative/terraform-provider-iterative/pull/632/commits/aaa06e9db2dd964870bc39749e8a8a609deb7c96 can pretty easily cherry-picked to another PR and rebased out?

Comment (tasdomas) [reply task/az/task.go:115]: Good point, will do just that - thank you!

**Review by 0x2b3bfa0 — COMMENTED** [review]
Is providing several user-assigned identities a common use case? 🤔

**Comment by dacbd** [comment]
I believe this is also funcationally unique to az, gcp/aws, barring common terminology, allow multiple policies to be attached to a single entity and one entity to be assigned to the instance.

**Review by 0x2b3bfa0 — COMMENTED** [review]
Merge #645 first?

**Review by 0x2b3bfa0 — COMMENTED** [review]
# Test

I'm getting the following error when performing the test described below, although it's probably just me doing something wrong or the `az` command-line tool behaving unexpectedly. Can you please take a look?

```
ERROR: Failed to connect to MSI. Please make sure MSI is configured correctly. 
Get Token request returned http error: 400, reason: Bad Request 
```

## Create a user-assigned identity

Write the code below to a file named `main.tf` and run `terraform init && terraform apply` to create the necessary resources:

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 2.98.0"
    }
  }
}

provider "azurerm" {
  features {}
}

data "azurerm_subscription" "current" {}

resource "azurerm_resource_group" "example" {
  name     = "example"
  location = "westus"
}

resource "azurerm_user_assigned_identity" "example" {
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location
  name                = "example"
}

resource "azurerm_role_assignment" "example" {
  principal_id       = azurerm_user_assigned_identity.example.principal_id
  role_definition_id = azurerm_role_definition.task.role_definition_resource_id
  scope              = data.azurerm_subscription.current.id
}

resource "azurerm_role_definition" "example" {
  name  = azurerm_user_assigned_identity.example.name
  scope = data.azurerm_subscription.current.id
  permissions {
    actions = ["Microsoft.Compute/virtualMachineScaleSets/read"]
  }
}

output "permission_set" {
  value = azurerm_user_assigned_identity.example.id
}
```

> **Note**
> Permissions in this role definition serve just as an example; users should customize them as they wish

## Create a test task using that identity

```bash
./terraform-provider-iterative create\
 --script='curl --location https://aka.ms/InstallAzureCLIDeb | bash && az login --identity && az vmss list'\
 --permission-set=$(terraform output --raw permission_set)\
 --region=westus\
 --cloud=az
```

> **Note**
> If running manually, you may want to replace `$(terraform output --raw permission_set)` with the permission set value, e.g. `/subscriptions/00000000-0000-0000-0000-000000000000
/resourceGroups/example/providers/Microsoft.ManagedIdentity/userAssignedIdentities/example`

---

Sorry for the late review. 🙈

**Comment by 0x2b3bfa0** [comment]
🔔  @tasdomas

**Comment by tasdomas** [comment]
> bell @tasdomas
Yeah - looking into this

**Comment by tasdomas** [comment]
The issue is two-fold here:
1. the command line permission set value was not being propagated to the task (fixed)
2. azure's permission system is complex - it's not enough to add a UMI with the necessary assigned roles (still working on this one)

**Comment by tasdomas** [comment]
> # Test
> 
> I'm getting the following error when performing the test described below, although it's probably just me doing something wrong or the `az` command-line tool behaving unexpectedly. Can you please take a look?
> 
> ```
> ERROR: Failed to connect to MSI. Please make sure MSI is configured correctly. 
> Get Token request returned http error: 400, reason: Bad Request 
> ```

So, this took a while to figure out, there were multiple issues:
1. the `permission_set` flag was not propagated to the task definition (fixed)
3. `az login --identity` will login with the machine's system-assigned identity by default (if it exists), the user-assigned identity needs to be explicitly specified to login with: `az login --identity -u {UMI-id}`. This is far from being intuitive, but we have limited ways of working around this **[1]**.


## [1] 

It appears the `az login --identity` logs in with the system assigned identity, if it exists, or the user assigned identity if only one is set. Currently, when allocating a machine, we maintain the system assigned identity for the scale set, so any user assigned identities need to be specifically logged-in with.

We could assign no system-managed identity if at least one UMI is specified. But this would make `permission_set` modify the allocation instead of augmenting it.

**Review by 0x2b3bfa0 — APPROVED** [review]
Updated https://github.com/iterative/terraform-provider-iterative/pull/632#pullrequestreview-1076595190 with working instructions; now it works as expected. 🎉

**Comment by 0x2b3bfa0** [comment]
With regard to the need of specifying an user I would suggest to simplify user's lives a bit more:
1. Don't accept a comma-separated list of identities; users will have to end up picking a single one to log in.
2. Expose the `permission_set` through an environment variable so users can run `az login --identity --username=$IDENTITY` or similar.

**Comment by 0x2b3bfa0** [comment]
> _we maintain the system assigned identity for the scale set_

What is it useful for? Maybe we can just go ahead and leave only the user-managed identity? E.g.

> _We could assign no system-managed identity if at least one UMI is specified. But this would make permission_set modify the allocation instead of augmenting it._

🤔

**Comment by 0x2b3bfa0** [comment]
Merging with (irrelevant) `aws` tests failing. Resources, resources... 😅
